### PR TITLE
Uses the closest point to calculate the equivalent p in a boundary.

### PR DESCRIPTION
**Pull request by francocipollone**
# 🦟 Bug fix

Related to #39 

## Summary

#39 Describes an issue that is evidenced when calculating the RBounds.
Intrinsically, this issue comes from the calculation of the equivalent p: this is the p in the left or right linestring that matches with the p in the centerline.

At the moment the equivalent p has been calculated by scaling the centerline p by the ratio of centerline length and boundary length. (Causing the issue in #39)

In this PR a new approach is given: **For a given point in the centerline, the equivalent p in the boundaries will be given by the closest point(boundary to centerline)**

Using this approach the issue of the RBounds calculation is solved, being consistent no matter the shape of the lane:

image
(Compare it with the case in #39 )

Pros:
 - This approach doesn't accumulate error as the prior does given that the closest point is calculated for each centerline p.
 - Solves the issue for the RBound calculation for any shape for roads.
 
Cons:
- The time for creating the meshes for the `s_shape_road` was doubled. From 36sec to 70sec(we should improve overall performance tho)
- when equivalent p is wanted to be calculated on a change of direction, an error might be added: this is minimized by sampling the road correctly and avoiding big discrete jumps in direction. For example:
   image

 

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if it affects the public API)

### CODE REVIEW

**Comment by francocipollone** [comment]
This is a draft and the tests will be updated once this approach has been debated.

**Comment by agalbachicar** [comment]
I think it is worth the pros against the cons because:

> - The time for creating the meshes for the s_shape_road was doubled. From 36sec to 70sec(we should improve overall performance tho)

This is a one time thing, and the problem should be reduced if we expose or use when computing meshes some kind of API that allows to pass in hints of where the previous value was computed, then the search is done in much smaller space. Also, sampling distance should reduce the computational cost here.

> when equivalent p is wanted to be calculated on a change of direction: this is minimized by sampling the road correctly and avoiding big discrete jumps in direction.

There is no formal solution to this problem, I think. It is heuristic based and we should define one way to do it. This one seems reasonable based on the results you showed.

File: src/geometry/lane_geometry.cc [hunk 1]
```diff
@@ -146,11 +146,11 @@ maliput::api::RBounds LaneGeometry::RBounds(double p) const {
...
   p = range_validator_(p);
   MALIPUT_THROW_UNLESS(line_string_type != LineStringType::kCenterLine);
-  // Given p in centerline, let's compute p_at_a_border as: p_at_a_border = p / length * length_at_a_border.
-  // See https://github.com/maliput/maliput_sparse/issues/15 for more details.
-  const double lateral_length = line_string_type == LineStringType::kLeftBoundary ? left_.length() : right_.length();
-  const double p_equivalent = p * lateral_length / centerline_.length();
-  return std::clamp(p_equivalent, 0., lateral_length);
+  const maliput::math::Vector3 point_at_p = utility::InterpolatedPointAtP(centerline_, p);
+  // The lateral p is the one that matches closest point in the lateral to the centerline at p.
+  const utility::ClosestPointResult closest_to_point = utility::GetClosestPointUsing2dProjection(
```
Comment (francocipollone) [inline src/geometry/lane_geometry.cc:151]: I am using a 2d projection for avoiding having incorrectly the closest point when the left and right linestring have variable differences in elevation along the p of the linestrings.

Imagine the case where:
 - left: {0.0 ,0.0, 0.0} -> {10.0 ,0.0, 10.0}
 - right: {0.0 ,-2.0, 0.0} -> {10.0 ,-2.0, 0.0}

**Comment by francocipollone** [comment]
While the issue is solved for the SShapeRoad (which is flat), an anomaly can be seen when there is a variable superelevation,  like in the SShapeSuperelevatedRoad. It is not as smooth as maliput_malidrive.

image

While in the part of the road where no superelevation is present.

image

**Comment by agalbachicar** [comment]
It is OK to not be smooth by definition. 
I think it is the combined effect between the curvature and the superelevation what makes that effect rather than superelevation alone.

File: include/maliput_sparse/geometry/utility/geometry.h [hunk 2]
```diff
@@ -38,23 +38,33 @@ namespace geometry {
 namespace utility {
 
 /// Holds the result of #GetBoundPointsAtP method.
+/// @tparam T The coordinate type.
+template <typename T>
```
Comment (agalbachicar) [inline include/maliput_sparse/geometry/utility/geometry.h:42]: I have a slight preference for CoordinateT instead of T. It is self-documented what it is expected to be.

File: test/geometry/lane_geometry_test.cc [hunk 3]
```diff
@@ -107,58 +107,81 @@ std::vector<OrientationTestCase> OrientationTestCases() {
...
+  virtual double GetTolerance() const {
+    static constexpr double kTolerance{1e-12};
+    return kTolerance;
+  }
+
+  void ExecuteOrientationTest() {
+    ASSERT_EQ(case_.p.size(), case_.expected_rpy.size()) << ">>>>> Test case is ill-formed.";
+    const LaneGeometry lane_geometry = case_.center.has_value()
+                                           ? LaneGeometry{case_.center.value(), case_.left, case_.right, 1e-3, 1.}
+                                           : LaneGeometry{case_.left, case_.right, 1e-3, 1.};
```
Comment (agalbachicar) [inline test/geometry/lane_geometry_test.cc:149]: Please, convert 1e-3 and 1. into named constants.

File: test/geometry/lane_geometry_test.cc [hunk 4]
```diff
@@ -342,26 +367,28 @@ std::vector<RBoundsCase> RBoundsCases() {
...
+              {-3.2015621187164243, 3.2015621187164243},
+              {-5.385164807134504, 5.385164807134504},
+          } /* expected_r_bounds */
       },
       {
           // Arc-like lane:
+          // For further information on this test case, see the following link:
+          // https://github.com/maliput/maliput_sparse/pull/46
           /*
-                  ______________
```
Comment (agalbachicar) [inline test/geometry/lane_geometry_test.cc:360]: Is this change on purpose?

**Review by agalbachicar — APPROVED** [review]
A few nits, it LGTM.

File: test/geometry/lane_geometry_test.cc (hunk 4 above)
Comment (francocipollone) [reply test/geometry/lane_geometry_test.cc:360]: Yes! It better represents the geometry of the lane.

File: test/geometry/lane_geometry_test.cc (hunk 3 above)
Comment (francocipollone) [reply test/geometry/lane_geometry_test.cc]: Done.

File: include/maliput_sparse/geometry/utility/geometry.h (hunk 2 above)
Comment (francocipollone) [reply include/maliput_sparse/geometry/utility/geometry.h]: Done.
